File: frontend/src/types/workflow.ts
export interface Position {
  x: number;
  y: number;
}

export type NodeType = 'user_input' | 'llm' | 'tts' | 'end';

export interface UserInputConfig {
  variable_name: string;
  placeholder: string;
}

export interface LLMConfig {
  model: string;
  api_base: string;
  api_key: string;
  system_prompt: string;
  prompt_template: string;
  temperature: number;
  max_tokens: number;
  top_p: number;
}

export interface TTSConfig {
  voice: string;
  speed: string;
  pitch: string;
  output_format: string;
}

export interface EndConfig {
  output_type: 'audio' | 'text' | 'auto';
}

export type NodeConfig = UserInputConfig | LLMConfig | TTSConfig | EndConfig;

export interface WorkflowNode {
  id: string;
  type: NodeType;
  label: string;
  position: Position;
  config: Record<string, unknown>;
}

export interface WorkflowEdge {
  id: string;
  source: string;
  target: string;
  sourceHandle: string;
  targetHandle: string;
}

export interface WorkflowDefinition {
  id: string;
  name: string;
  description: string;
  nodes: WorkflowNode[];
  edges: WorkflowEdge[];
  created_at: string;
  updated_at: string;
}

export interface WorkflowSummary {
  id: string;
  name: string;
  description: string;
  node_count: number;
  updated_at: string;
}

File: backend/models/workflow.py
from __future__ import annotations

from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, Field


class NodeType(str, Enum):
    USER_INPUT = "user_input"
    LLM = "llm"
    TTS = "tts"
    END = "end"


class Position(BaseModel):
    x: float = 0
    y: float = 0


class UserInputConfig(BaseModel):
    variable_name: str = "user_input"
    placeholder: str = "Please enter text..."


class LLMConfig(BaseModel):
    model: str = "gpt-3.5-turbo"
    api_base: str = ""
    api_key: str = ""
    system_prompt: str = ""
    prompt_template: str = "{{user_input}}"
    temperature: float = 0.7
    max_tokens: int = 2048
    top_p: float = 1.0


class TTSConfig(BaseModel):
    voice: str = "zh-CN-XiaoxiaoNeural"
    speed: str = "+0%"
    pitch: str = "+0Hz"
    output_format: str = "mp3"


class EndConfig(BaseModel):
    output_type: str = "auto"


class WorkflowNode(BaseModel):
    id: str
    type: NodeType
    label: str = ""
    position: Position = Field(default_factory=Position)
    config: dict[str, Any] = Field(default_factory=dict)


class WorkflowEdge(BaseModel):
    id: str
    source: str
    target: str
    sourceHandle: str = "output"
    targetHandle: str = "input"


class WorkflowDefinition(BaseModel):
    id: str = ""
    name: str = "Untitled Workflow"
    description: str = ""
    nodes: list[WorkflowNode] = Field(default_factory=list)
    edges: list[WorkflowEdge] = Field(default_factory=list)
    created_at: str = ""
    updated_at: str = ""


class WorkflowSummary(BaseModel):
    id: str
    name: str
    description: str
    node_count: int
    updated_at: str
